### PR TITLE
Add handling added/deleted files in diff

### LINKED ISSUE #49: MultiFileDiffReader doesn’t handle messages in diff that certain files are only available in specific version.
In some unified diff files there are such lines as `Only in {path}: {filename}`, meaning that certain files are only available in specific version. Those messages aren’t handled in any way, except when those lines are just catched in FileDiff.Extended for next FileDiff in the file.
	Are there any plans of handling these messages?

For example, those unique files can be picked up in next functions, representing them as FileDiff with only the value OrigName.
```
func (r *MultiFileDiffReader) ReadAllFiles() ([]*FileDiff, error)
func (r *MultiFileDiffReader) ReadFile() (*FileDiff, error)
```

Then for this file example, the output below for ReadAllFiles() would be expected.
my_diff.txt
```diff
diff -u source_a/file_1.txt  source_b/file_1.txt
--- source_a/file_1.txt   2020-07-28 12:54:18.000000000 +0000
+++ source_b/file_1.txt  2020-07-28 12:54:18.000000000 +0000
@@ -2,3 +3,4 @@
 To be, or not to be, that is the question:
-Whether 'tis nobler in the mind to suffer
+The slings and arrows of outrageous fortune,
+Or to take arms against a sea of troubles
 And by opposing end them. To die—to sleep,
Only in source_a: file_2.txt
Only in source_b: file_3.txt
```
```
ReadAllFiles(myMultiFileDiffReader) -> {
    FileDiff{
        OrigName: “source_a/file_1.txt”, 
        OrigTime: ..., 
        NewName: “source_b/file_1.txt”, 
        NewTime: …,
        Entended: ..., 
        Hunks: …
    },
 
    FileDiff{
        OrigName: “source_a/file_2.txt”, 
        OrigTime: nil, 
        NewName: nil, 
        NewTime: nil
        …
    }, 

    FileDiff{
        OrigName: “source_b/file_3.txt”, 
        OrigTime: nil, 
        NewName: nil, 
        …
    }

}, nil
```

### PR DESCRIPTION
**Pull request by sofiia-tesliuk**
Related issue: https://github.com/sourcegraph/go-diff/issues/49
closes #49 

There are some changes from planned implementation:
- empty NewName marked as empty string instead of `nil`
- files aren't determined if they are from original source or new one. All filenames from such messages will be added to `OrigName` field.
 `[2].OrigName: "source_b/file_3.txt` and `[2].NewName: ""` will be relevant.
Because in case diff file only contains messages about added/deleted files, it is impossible to determine which of the sources is the original one.

Then for this file example, the output below for ReadAllFiles() would be expected.
my_diff.txt
```diff
diff -u source_a/file_1.txt  source_b/file_1.txt
--- source_a/file_1.txt   2020-07-28 12:54:18.000000000 +0000
+++ source_b/file_1.txt  2020-07-28 12:54:18.000000000 +0000
@@ -2,3 +3,4 @@
 To be, or not to be, that is the question:
-Whether 'tis nobler in the mind to suffer
+The slings and arrows of outrageous fortune,
+Or to take arms against a sea of troubles
 And by opposing end them. To die—to sleep,
Only in source_a: file_2.txt
Only in source_b: file_3.txt
```
```
ReadAllFiles(myMultiFileDiffReader) -> {
    FileDiff{
        OrigName: “source_a/file_1.txt”, 
        OrigTime: ..., 
        NewName: “source_b/file_1.txt”, 
        NewTime: …,
        Entended: ..., 
        Hunks: …
    },
 
    FileDiff{
        OrigName: “source_a/file_2.txt”, 
        OrigTime: nil, 
        NewName: "", 
        NewTime: nil
        …
    }, 

    FileDiff{
        OrigName: “source_b/file_3.txt”, 
        OrigTime: nil, 
        NewName: "", 
        …
    }

}, nil
```

### CODE REVIEW

File: diff/diff_test.go [hunk 1]
```diff
@@ -496,6 +496,60 @@ func TestParseMultiFileDiffHeaders(t *testing.T) {
...
+				},
+				{
+					OrigName: "source_a/file_2.txt",
+					OrigTime: nil,
+					NewName:  "",
+					NewTime:  nil,
+					Extended: nil,
+				},
+				{
+					OrigName: "source_b/file_3.txt",
```
Comment (schachmat) [inline diff/diff_test.go:545]: maybe make one of the 3 files be added, not all of them deleted? Or split the test case into two:
* One where files are only added
* One where files are only deleted

File: diff/parse.go [hunk 2]
```diff
@@ -72,6 +73,12 @@ func (r *MultiFileDiffReader) ReadFile() (*FileDiff, error) {
 		}
 	}
 
+	// FileDiff is added/deleted file
+	// No further collection of hunks needed
+	if fd.NewName == "" {
```
Comment (schachmat) [inline diff/parse.go:78]:
```suggestion
	if fd.NewName == "" || fd.OldName == "" {
```

Doesn't this need to check for the old file name too to catch the case where we're creating/adding a file?

File: diff/parse.go [hunk 3]
```diff
@@ -223,8 +230,31 @@ func (r *FileDiffReader) HunksReader() *HunksReader {
 
 // ReadFileHeaders reads the unified file diff header (the lines that
 // start with "---" and "+++" with the orig/new file names and
-// timestamps).
+// timestamps). Or starts with "Only in " message with dir path filename.
 func (r *FileDiffReader) ReadFileHeaders() (origName, newName string, origTimestamp, newTimestamp *time.Time, err error) {
+	if (r.fileHeaderLine != nil) && (bytes.HasPrefix(r.fileHeaderLine, onlyInMessagePrefix)) {
+		path := bytes.Split(bytes.TrimPrefix(
```
Comment (schachmat) [inline diff/parse.go:236]: I think regexp matching with Submatches would be much clearer here. Alternatively you can try one of the fmt.Scanf functions.

File: diff/parse.go [hunk 4]
```diff
@@ -223,8 +230,31 @@ func (r *FileDiffReader) HunksReader() *HunksReader {
...
+			bytes.TrimSuffix(r.fileHeaderLine, []byte("\n")),
+			onlyInMessagePrefix), []byte(":"))
+		if len(path) != 2 {
+			return "", "", nil, nil,
+				&ParseError{r.line, r.offset, ErrBadOnlyInMessage}
```
Comment (schachmat) [inline diff/parse.go:241]: Not sure if we want to throw an error here. Is it not allowed to have an extended header line that starts with "Only in ", but is not about an added/removed file?

File: diff/parse.go [hunk 5]
```diff
@@ -223,8 +230,31 @@ func (r *FileDiffReader) HunksReader() *HunksReader {
...
+		unquotedSource, err := strconv.Unquote(source)
+		if err == nil {
+			source = unquotedSource
+		}
+		unquotedFilename, err := strconv.Unquote(filename)
+		if err == nil {
+			filename = unquotedFilename
+		}
+
+		return filepath.Join(source, filename), "", nil, nil, nil
```
Comment (schachmat) [inline diff/parse.go:255]: Do we not have to interpret `source` to figure out if `filename` was added or removed and depending on that `return "", filepath.Join(source, filename), nil, nil, nil` instead?

File: diff/parse.go [hunk 6]
```diff
@@ -324,7 +354,7 @@ func (r *FileDiffReader) ReadExtendedHeaders() ([]string, error) {
 				return xheaders, OverflowError(line)
 			}
 		}
-		if bytes.HasPrefix(line, []byte("--- ")) {
+		if bytes.HasPrefix(line, []byte("--- ")) || (bytes.HasPrefix(line, onlyInMessagePrefix)) {
```
Comment (schachmat) [inline diff/parse.go:357]: Similar to above. Is it not allowed to have an extended header line starting with "Only in "? Maybe even an extended header before the actual "Only in " line?

File: diff/parse.go [hunk 7]
```diff
@@ -403,6 +433,10 @@ var (
 
 	// ErrExtendedHeadersEOF is when an EOF was encountered while reading extended file headers, which means that there were no ---/+++ headers encountered before hunks (if any) began.
 	ErrExtendedHeadersEOF = errors.New("expected file header while reading extended headers, got EOF")
+
+	// ErrBadOnlyInMessage is when a file have a malformed `only in` message
+	// Should be in format `Only in {source}: {filename}`
+	ErrBadOnlyInMessage = errors.New("bad `only in` message")
```
Comment (schachmat) [inline diff/parse.go:439]: I'd avoid using backticks in the error message. Can you use `"` or `'` instead?

File: diff/print.go [hunk 8]
```diff
@@ -36,6 +37,16 @@ func PrintFileDiff(d *FileDiff) ([]byte, error) {
 		}
 	}
 
+	// FileDiff is added/deleted file
+	// No further hunks printing needed
+	if d.NewName == "" {
```
Comment (schachmat) [inline diff/print.go:42]: If you do the changes as suggested in parse.go, then you probably need to update this here too to also catch `if d.OldName == ""`.

File: diff/diff_test.go (hunk 1 above)
Comment (sofiia-tesliuk) [reply diff/diff_test.go:545]: Please see PR description. All filenames from `Only in {source}: {filename}` messages will be added to `OrigName` field.

File: diff/parse.go (hunk 2 above)
Comment (sofiia-tesliuk) [reply diff/parse.go:79]: All filenames from `Only in {source}: {filename}` messages will be added to `OrigName` field.

File: diff/parse.go (hunk 5 above)
Comment (sofiia-tesliuk) [reply diff/parse.go]: All filenames from `Only in {source}: {filename}` messages will be added to `OrigName` field.

File: diff/print.go (hunk 8 above)
Comment (sofiia-tesliuk) [reply diff/print.go:42]: All filenames from `Only in {source}: {filename}` messages will be added to `OrigName` field.

File: diff/parse.go [hunk 9]
```diff
@@ -612,6 +637,18 @@ func (r *HunksReader) ReadAllHunks() ([]*Hunk, error) {
...
+// parseOnlyInMessage checks if line is a "Only in {source}: {filename}" and returns source and filename
+func parseOnlyInMessage(line []byte) (bool, []byte, []byte) {
+	re := regexp.MustCompile("Only in ([\x21-\x7E]+): ([\x21-\x7E]+)")
+	if re.Match(line) {
+		slices := re.FindSubmatch(line)
+		if fmt.Sprintf("Only in %s: %s", slices[1], slices[2]) == string(line) {
+			return true, slices[1], slices[2]
+		}
+	}
+	return false, nil, nil
```
Comment (keegancsmith) [inline diff/parse.go:649]: FYI I did some testing, and it seems the output of BSD diff doesn't do any escaping/etc: For example if you have a file or dir containing `: ` in it, it will just be output verbatim. This makes it impossible to always reconstruct this. For example look at this fun output I constructed:

```shellsession
$ ls internal/trace/foo\ bar internal/trace/hello:\ world
internal/trace/foo bar:
bam      bam: bar

internal/trace/hello: world:
bazz

$ diff -r internal/trace/foo\ bar internal/trace/hello:\ world
Only in internal/trace/foo bar: bam
Only in internal/trace/foo bar: bam: bar
Only in internal/trace/hello: world: bazz
```

This makes me think the regex is too restrictive. I'd suggest using more straightforward non-regex code:
```suggestion
	if !bytes.HasPrefix(line, onlyInMessagePrefix) {
		return false, nil, nil
	}
	line = line[len(onlyInMessagePrefix):]
	idx := bytes.Index(line, []byte(": "))
	if idx < 0 {
		return false, nil, nil
	}
	return true, line[:idx], line[idx+2:]
```

What do you think?

File: diff/testdata/sample_onlyin_line_isnt_a_file_header.diff [hunk 10]
```diff
@@ -0,0 +1,13 @@
...
++++ source_b/file_1.txt
+@@ -2,3 +3,4 @@
+ To be, or not to be, that is the question:
+-Whether 'tis nobler in the mind to suffer
++The slings and arrows of outrageous fortune,
++Or to take arms against a sea of troubles
+ And by opposing end them. To die—to sleep,
+Only in universe!
+Only in source_a: file_2.txt
+Only in source_b: file_3.txt some unrelated stuff here.
```
Comment (keegancsmith) [inline diff/testdata/sample_onlyin_line_isnt_a_file_header.diff:12]: when does "some unrelated stuff" happen? That runs counter to my suggestion above around just splitting on `: `, which means a more restrictive regex makes sense.

**Comment by keegancsmith** [comment]
BTW thank you so much for the great pull request!

File: diff/parse.go [hunk 11]
```diff
@@ -223,8 +231,15 @@ func (r *FileDiffReader) HunksReader() *HunksReader {
 
 // ReadFileHeaders reads the unified file diff header (the lines that
 // start with "---" and "+++" with the orig/new file names and
-// timestamps).
+// timestamps). Or starts with "Only in " message with dir path filename.
```
[the remark this answers is not in the document]
Comment (sofiia-tesliuk) [reply diff/parse.go:234]: One more message that will be relevant to handle is `Common subdirectories: %s and %s\n`, which actually doesn't notify about change. Also will be quite hard to fit it into `FileDiff` as it is a message about directories. Maybe it is better to create an issue about it and think what to do with this message later.

File: diff/parse.go (hunk 9 above)
Comment (sofiia-tesliuk) [reply diff/parse.go:649]: I think that situation with complex filename (containing `: `) is much more common than situation when line, which is not a part of the hunk, have passed straightforward non-regex check, but isn't actually an `Only in` message.
So far it probably will be better to have not that strict check.
Have added a test case with your example :)

File: diff/testdata/sample_onlyin_line_isnt_a_file_header.diff (hunk 10 above)
Comment (sofiia-tesliuk) [reply diff/testdata/sample_onlyin_line_isnt_a_file_header.diff:12]: Just example that there can be smth in the same line after `Only in` message. But considering not strict check, this `some unrelated stuff here` will be a part of filename.

File: diff/parse.go [hunk 12]
```diff
@@ -612,6 +637,18 @@ func (r *HunksReader) ReadAllHunks() ([]*Hunk, error) {
 	}
 }
 
+// parseOnlyInMessage checks if line is a "Only in {source}: {filename}" and returns source and filename
+func parseOnlyInMessage(line []byte) (bool, []byte, []byte) {
+	re := regexp.MustCompile("Only in ([\x21-\x7E]+): ([\x21-\x7E]+)")
```
[the remark this answers is not in the document]
Comment (sofiia-tesliuk) [reply diff/parse.go:642]: Changed to non-regex check for now.

File: diff/parse.go (hunk 11 above)
[the remark this answers is not in the document]
Comment (keegancsmith) [reply diff/parse.go]: Yeah sorry I read my message again and I see that it implies you should support it in this PR. No need to do that, that would be out of scope for a single PR.

File: diff/parse.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (schachmat) [reply diff/parse.go]: I think the comment was about having the regexp as a package-scoped var like this:

```
var (
  reOnlyIn := regexp.MustCompile("Only in ([\x21-\x7E]+): ([\x21-\x7E]+)")
)
```

Then the regexp is compiled at load time of the library and if there is an issue it fails right away rather than in the middle of program execution.

[the remark this answers is not in the document]
Comment (sofiia-tesliuk) [reply diff/parse.go]: Yes, but in the comment below there was suggested a change to use less strict message check without regex.
Please see the updated version of [`parseOnlyInMessage`](https://github.com/sofiia-tesliuk/go-diff/blob/29fa046e33bb9ccb5992240fe27380b709e750d0/diff/parse.go#L641).

File: diff/parse.go (hunk 9 above)
Comment (schachmat) [reply diff/parse.go]: Can we just use this regexp? `^Only in ([^:]+): (.+)$`

This should be simple enough and avoid the problem mentioned above, right?
